pyhap/accessory_driver.py

```diff
@@ -859,118 +860,106 @@ def set_characteristics(self, chars_query, client_addr):
 
-            if HAP_PERMISSION_NOTIFY in cq:
-                char_topic = get_topic(aid, iid)
-                action = "Subscribed" if cq[HAP_PERMISSION_NOTIFY] else "Unsubscribed"
-                logger.debug(
-                    "%s client %s to topic %s", action, client_addr, char_topic
-                )
-                self.async_subscribe_client_topic(
-                    client_addr, char_topic, cq[HAP_PERMISSION_NOTIFY]
-                )
+        primary_accessory = self.accessory
+        primary_aid = primary_accessory.aid
 
-            if HAP_REPR_VALUE not in cq:
+        for query in queries:
+            if HAP_REPR_VALUE not in query and not expired:
                 continue
 
-            updates.setdefault(aid, {})[iid] = cq[HAP_REPR_VALUE]
+            aid, iid = query[HAP_REPR_AID], query[HAP_REPR_IID]
+            value = query.get(HAP_REPR_VALUE, None)
+            write_response_requested = query.get(HAP_REPR_WRITE_RESPONSE, False)
 
-        for aid, new_iid_values in updates.items():
-            if self.accessory.aid == aid:
-                acc = self.accessory
+            if aid == primary_aid:
+                acc = primary_accessory
             else:
                 acc = self.accessory.accessories.get(aid)
+            char = acc.get_characteristic(aid, iid)
 
-            updates_by_service = {}
-            char_to_iid = {}
-            for iid, value in new_iid_values.items():
-                # Characteristic level setter callbacks
-                char = acc.get_characteristic(aid, iid)
+            set_result, set_result_value = (
+                HAP_SERVER_STATUS.INVALID_VALUE_IN_REQUEST,
+                None,
+            )
+            if value is not None:
+                set_result, set_result_value = _wrap_char_setter(
+                    char, value, client_addr
+                )
 
-                set_result, set_result_value = _wrap_char_setter(char, value, client_addr)
-                if set_result != HAP_SERVER_STATUS.SUCCESS:
-                    had_error = True
+            return_value_in_response = (
+                set_result_value is not None and write_response_requested
+            )
 
-                setter_results[aid][iid] = set_result
+            if aid not in results:
+                results[aid] = {}
+            if iid not in results[aid]:
+                results[aid][iid] = {}
 
-                if set_result_value is not None:
-                    if setter_responses.get(aid, None) is None:
-                        logger.warning(
-                            "Returning write response '%s' when it wasn't requested for %s %s",
-                            set_result_value, aid, iid
-                        )
-                    had_write_response = True
-                    setter_responses.setdefault(aid, {})[iid] = set_result_value
+            results[aid][iid] = {HAP_REPR_STATUS: set_result}
```

Comment:
I would assign the new dict to a local var and than set results[aid][iid] = var at the end of the loop

This avoids all the lookups to find the inner dict each time you set a key/value pair